### PR TITLE
feat(profiling): Introduce use case column for DSNs

### PR DESCRIPTION
**Pull request by jjbayer**
For features that need to emit metrics from the sentry code base, such as [Aggregated Functions](https://www.notion.so/sentry/Moving-Profile-Metric-to-Generic-Metrics-Platform-194ead02a6a84b0b81939c7ef88e9fef) and [Escalating Issues](https://github.com/getsentry/sentry/blob/e7fae9c02e49f18f9ab40dedb8bd36701bfdea22/src/sentry/tasks/post_process.py#L146), we need the possibility to send metrics without associating them to a specific user DSN.

ref: https://github.com/getsentry/relay/issues/3114

### CODE REVIEW

File: src/sentry/models/projectkey.py [hunk 1]
```diff
@@ -89,13 +99,16 @@ class roles(TypedClassBitField):
...
-        cache_ttl=60 * 30,
     )
 
     data: models.Field[dict[str, Any], dict[str, Any]] = JSONField()
 
+    use_case = models.CharField(
+        max_length=32,
+        choices=[(v.value, v.value) for v in UseCase],
+        default=UseCase.USER.value,
+    )
```
Comment (jjbayer) [inline src/sentry/models/projectkey.py:110]: I considered encoding the use case into the existing `roles` field. But `roles` specify the permissions a DSN has, while the use case defines who may use the DSN.

File: src/sentry/api/serializers/models/project_key.py [hunk 2]
```diff
@@ -100,4 +103,8 @@ def serialize(
                 "hasDebug": get_dynamic_sdk_loader_option(obj, DynamicSdkLoaderOption.HAS_DEBUG),
             },
         }
+
+        if user.is_superuser:
```
Comment (markstory) [inline src/sentry/api/serializers/models/project_key.py:107]: Superuser is being split into two roles (staff and superuser). Staff role users will have access to internal tools like `_admin` but not to customer data. How does superuser create these internal DSNs?

File: tests/sentry/api/endpoints/test_project_keys.py [hunk 3]
```diff
@@ -20,6 +24,48 @@ def test_simple(self):
...
+        url = reverse(
+            "sentry-api-0-project-keys",
+            kwargs={"organization_slug": project.organization.slug, "project_slug": project.slug},
+        )
+        response = self.client.get(url)
+        assert response.status_code == 200
+        assert len(response.data) == 1
+        response_data = response.data[0]
+        assert "use_case" not in response_data
+        assert response_data["public"] == key1.public_key
```
Comment (markstory) [inline tests/sentry/api/endpoints/test_project_keys.py:41]: The linked design doc states that internal DSNs should not show up in public APIs. Should this be a 404 as the user doesn't have an active superuser/staff session.

File: src/sentry/api/endpoints/project_keys.py [hunk 4]
```diff
@@ -51,9 +51,8 @@ def get(self, request: Request, project) -> Response:
         """
         Return a list of client keys bound to a project.
         """
-        queryset = ProjectKey.objects.filter(
-            project=project, roles=F("roles").bitor(ProjectKey.roles.store)
-        )
+        objects = ProjectKey.objects if request.user.is_superuser else ProjectKey.objects.user()
```
Comment (markstory) [inline src/sentry/api/endpoints/project_keys.py:54]: Usually we also check that the user has an active superuser/staff session as well.

File: src/sentry/api/serializers/models/project_key.py (hunk 2 above)
Comment (jjbayer) [reply src/sentry/api/serializers/models/project_key.py:107]: The intention here is that sentry developers can view the internal DSNs of a project, but the customers themselves cannot. These internal DSNs are not really customer data, but they're associated to a customer project. Would you say `staff` would be the more appropriate role for this?

File: tests/sentry/api/endpoints/test_project_keys.py (hunk 3 above)
Comment (jjbayer) [reply tests/sentry/api/endpoints/test_project_keys.py]: I wanted to test that regular users can see user keys but not internal keys in the list endpoint. I added a comment and better variable names to clarify.

File: src/sentry/api/endpoints/project_key_details.py [hunk 5]
```diff
@@ -56,8 +56,9 @@ def get(self, request: Request, project, key_id) -> Response:
         """
         Return a client key bound to a project.
         """
+        objects = ProjectKey.objects if request.user.is_superuser else ProjectKey.objects.user()
         try:
-            key = ProjectKey.objects.get(
+            key = objects.user().get(
```
Comment (iker-barriocanal) [inline src/sentry/api/endpoints/project_key_details.py:61]:
```suggestion
            key = objects.get(
```

File: tests/sentry/api/endpoints/test_project_key_details.py [hunk 6]
```diff
@@ -287,3 +311,20 @@ def test_simple(self):
         resp = self.client.delete(url)
         assert resp.status_code == 204, resp.content
         assert not ProjectKey.objects.filter(id=key.id).exists()
+
+    def test_use_case(self):
+        """Cannot delete an internal DSN"""
```
Comment (iker-barriocanal) [inline tests/sentry/api/endpoints/test_project_key_details.py:316]: What's the process if we need to delete/rotate an internal DSN?

File: src/sentry/models/projectkey.py [hunk 7]
```diff
@@ -51,6 +52,19 @@ def post_delete(self, instance, **kwargs):
...
+        return self.get_queryset().filter(use_case=UseCase.USER.value)
+
+
+class UseCase(enum.Enum):
+    # A user-visible project key.
+    USER = "user"
+    # An internal project key for submitting aggregate function metrics.
+    PROFILING = "profiling"
+    # An internal project key for submitting escalating issues metrics.
+    ESCALATING_ISSUES = "escalating_issues"
```
Comment (iker-barriocanal) [inline src/sentry/models/projectkey.py:66]: I think enum variants are documented like this in Python (I didn't find docs/PEP in some quick searches):
```suggestion
    USER = "user"
    """A user-visible project key."""
    PROFILING = "profiling"
    """An internal project key for submitting aggregate function metrics."""
    ESCALATING_ISSUES = "escalating_issues"
    """An internal project key for submitting escalating issues metrics."""
```

**Review by iker-barriocanal — COMMENTED** [review]
It LGTM, but I don't have a lot of insights around this part of the codebase so don't rely on my review 😄

**Review by Dav1dde — COMMENTED** [review]
Changes look good to me, but I have no overview whether there is more that needs to be updated.

File: src/sentry/api/serializers/models/project_key.py (hunk 2 above)
Comment (schew2381) [reply src/sentry/api/serializers/models/project_key.py:107]: In that case, the check can remain as superuser. Mark also mentioned this but to check for active superuser you can use `is_active_superuser`.

`user.superuser` is only checking that a user can become an active superuser, not that they actually are one.

Comment (jjbayer) [reply src/sentry/api/serializers/models/project_key.py:107]: @schew2381 thanks! Replaced with `is_active_superuser` check now.

File: src/sentry/api/endpoints/project_key_details.py (hunk 5 above)
Comment (jjbayer) [reply src/sentry/api/endpoints/project_key_details.py]: Thanks, good catch!

File: src/sentry/api/endpoints/project_keys.py (hunk 4 above)
Comment (jjbayer) [reply src/sentry/api/endpoints/project_keys.py]: Thanks, changed to `is_active_superuser` now.

File: src/sentry/models/projectkey.py (hunk 7 above)
Comment (jjbayer) [reply src/sentry/models/projectkey.py]: Makes sense, updated.

File: tests/sentry/api/endpoints/test_project_key_details.py (hunk 6 above)
Comment (jjbayer) [reply tests/sentry/api/endpoints/test_project_key_details.py]: Super should be allowed to update / delete a key. Added a test for that now.

File: src/sentry/api/serializers/models/project_key.py (hunk 2 above)
Comment (markstory) [inline src/sentry/api/serializers/models/project_key.py:107]: Should this be `is_active_superuser` as well?

File: src/sentry/services/hybrid_cloud/project_key/impl.py [hunk 8]
```diff
@@ -14,7 +14,7 @@ def _get_project_key(self, project_id: str, role: ProjectKeyRole) -> RpcProjectK
         from sentry.models.projectkey import ProjectKey
 
         project_keys = ProjectKey.objects.filter(
-            project=project_id, roles=F("roles").bitor(role.as_orm_role())
+            use_case="user", project=project_id, roles=F("roles").bitor(role.as_orm_role())
```
Comment (iker-barriocanal) [inline src/sentry/services/hybrid_cloud/project_key/impl.py:17]: Can we reuse the enum in these calls?

**Review by iker-barriocanal — APPROVED** [review]
LGTM, but you may want the approval of somebody with expertise in this area 😄

File: src/sentry/services/hybrid_cloud/project_key/impl.py (hunk 8 above)
Comment (markstory) [reply src/sentry/services/hybrid_cloud/project_key/impl.py:17]: Yes, enum values could be used here.

**Review by markstory — APPROVED** [review]
Schema migration looks good.

File: src/sentry/services/hybrid_cloud/project_key/impl.py (hunk 8 above)
Comment (jjbayer) [reply src/sentry/services/hybrid_cloud/project_key/impl.py]: 👍 using enum everywhere now.

File: src/sentry/api/serializers/models/project_key.py (hunk 2 above)
Comment (jjbayer) [reply src/sentry/api/serializers/models/project_key.py]: Thanks, passed in the request to the serializer now to call `is_active_superuser`.

**Pull request merged**